Source/WebCore/page/FrameView.cpp

```diff
@@ -2446,13 +2447,29 @@ void FrameView::scrollToFocusedElementInternal()
     if (!updateTarget)
         return;
 
+    // Get the scroll-margin of the shadow host when we're inside a user agent shadow root.
+    if (updateTarget->containingShadowRoot() && updateTarget->containingShadowRoot()->mode() == ShadowRootMode::UserAgent)
+        updateTarget = updateTarget->shadowHost();
+
     auto* renderer = updateTarget->renderer();
     if (!renderer || renderer->isWidget())
         return;
 
+    auto writingMode = renderer->style().writingMode();
```

Comment:
This is repeating code that exists in `Element::scrollIntoView()`. Can we share it?